aws-s3-transfer-manager/src/operation/upload_objects/worker.rs

```diff
@@ -0,0 +1,507 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::borrow::Cow;
+use std::path::MAIN_SEPARATOR;
+use std::sync::atomic::Ordering;
+
+use super::{UploadObjectsContext, UploadObjectsInput};
+use async_channel::{Receiver, Sender};
+use aws_smithy_types::error::operation::BuildError;
+use blocking::Unblock;
+use futures_util::StreamExt;
+use walkdir::WalkDir;
+
+use crate::io::InputStream;
+use crate::operation::upload::UploadInputBuilder;
+use crate::operation::DEFAULT_DELIMITER;
+use crate::types::{FailedTransferPolicy, FailedUploadTransfer, UploadFilter};
+use crate::{error, types::UploadFilterItem};
+
+#[derive(Debug)]
+pub(super) struct UploadObjectJob {
+    key: String,
+    object: InputStream,
+}
+
+impl UploadObjectJob {
+    fn new(key: String, object: InputStream) -> Self {
+        Self { key, object }
+    }
+}
+
+pub(super) async fn list_directory_contents(
+    input: UploadObjectsInput,
+    work_tx: Sender<Result<UploadObjectJob, error::Error>>,
+) -> Result<(), error::Error> {
+    // Move a blocking I/O to a dedicated thread pool
+    let mut walker = Unblock::new(walker(&input).into_iter());
```

Comment:
Would it be simpler to spawn this whole `list_directory_contents()` task via tokio's [spawn_blocking()](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html)?

Reply:
Still don't think it's worth bringing in another 3rdparty crate, with its own blocking thread pool, when tokio has already has a blocking thread pool

Now that I understand blocking a bit better ... it's just the `walker.next()` call we'd wrap with tokio's spawn_blocking. We wouldn't wrap the whole list_directory_contents task.

Reply:
I don't know that we want to call `spawn_blocking` on every `walker.next()` call. That's a potential context switch for every time we fetch the next entry. You'd want to wrap the entire iterator into a `spawn_blocking` task that sends results over a channel if you go that route. I'm indifferent on the 3p crate addition at the moment, until I see performance numbers it's all just something internal that can be refactored.

Reply:
my (extremely shallow) understand of this Unblock crate is that it's doing the exact same thing: a context switch while it runs `walker.next()` in blocking thread pool. Except it's an extra dependency and an extra blocking-thread-pool.

my pushback wasn't from a performance standpoint, just from a extra dependency/complexity standpoint

but yeah, we can refactor it out later

Reply:
Makes sense. I realized, however, that switching to use `spawn_blocking` brings a few implications. The most notable one is that we'd have to pass non-async function to `spawn_blocking` but sending directory traversal results currently involves an async channel. So there would be some impedance mismatch between sync and async.

Since use of the `blocking` crate is internal, I left `TODO` in 245e47b and kept the code as is for now.